# ch15 Keeping Time, Scheduling Tasks, and Launching Programs

## Keeping Time, Scheduling Tasks, and Launching Programs

- 내가 꼭 컴퓨터 앞에 앉아 있어야 하는가? 라는 고찰
- 내가 컴퓨터 앞에 없어도 시간이 되면 알아서 실행이 되고 알아서 결과고 메일로 날라오는 자동화 시스템이 내가 꿈꾸는 것
- time, datetime 모듈이 이러한 기능들을 지원해준다.

## The time Module

- time.time()
- time.sleep()

### The time.time() Function

- Unix epoch
- 12 AM on January 1, 1970
- UTC(Universal Time)
- [협정 세계시 \- 위키백과, 우리 모두의 백과사전](http://ko.wikipedia.org/wiki/협정_세계시): UTC - Coordinated Universal Time: 왜 거꾸로 됐는가?
- [Coordinated Universal Time \- Wikipedia, the free encyclopedia](http://en.wikipedia.org/wiki/Coordinated_Universal_Time)

In [1]:
import time

In [4]:
time.time()

1433658332.146624

In [5]:
time.time()

1433658333.241262

In [9]:
import time

def calcProd():
    # Calculate the product of the first 100,000 numbers.
    product = 1
    for i in range(1, 100000):
        product *= i
    return product

startTime = time.time()
prod = calcProd()
endTime = time.time()

print('The result is {} digits long.'.format(len(str(prod))))
print('Took {} seconds to calculate.'.format(endTime - startTime))

The result is 456569 digits long.
Took 3.18268203735 seconds to calculate.


- [27.4. The Python Profilers — Python 3.4.3 documentation](https://docs.python.org/3/library/profile.html): profiler가 더 좋은 방법

In [10]:
%%writefile profile_calc.py
def calcProd():
    # Calculate the product of the first 100,000 numbers.
    product = 1
    for i in range(1, 100000):
        product *= i
    return product

prod = calcProd()

print('The result is {} digits long.'.format(len(str(prod))))

Writing profile_calc.py


In [11]:
!python -m cProfile profile_calc.py

The result is 456569 digits long.
         6 function calls in 6.085 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    2.863    2.863    6.085    6.085 profile_calc.py:1(<module>)
        1    3.220    3.220    3.222    3.222 profile_calc.py:1(calcProd)
        1    0.000    0.000    0.000    0.000 {len}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 {method 'format' of 'str' objects}
        1    0.002    0.002    0.002    0.002 {range}




In [12]:
!python -m cProfile profile_calc.py

The result is 456569 digits long.
         6 function calls in 6.075 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    2.866    2.866    6.075    6.075 profile_calc.py:1(<module>)
        1    3.207    3.207    3.209    3.209 profile_calc.py:1(calcProd)
        1    0.000    0.000    0.000    0.000 {len}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 {method 'format' of 'str' objects}
        1    0.002    0.002    0.002    0.002 {range}




### The time.sleep() Function

In [13]:
import time

In [14]:
for i in range(3):
    print('Tick')
    time.sleep(1)
    print('Tock')
    time.sleep(1)

Tick
Tock
Tick
Tock
Tick
Tock


## Rounding Numbers

In [15]:
import time

In [16]:
now = time.time()

In [17]:
now

1433659402.314365

In [18]:
round(now, 2)

1433659402.31

In [19]:
round(now, 4)

1433659402.3144

In [20]:
round(now)

1433659402.0

## Project: Super Stopwatch

### High level Logic

- Track the amount of time elapsed between presses of the ENTER key, with each key press starting a new "lap" on the timer.
- Print the lap number, total time, and lap time.

### Code level Logic

- Find the current time by calling time.time() and store it as a timestamp at the start of the program, as well as at the start of each lap.
- Keep a lap counter and increment it every time the user presses ENTER.
- Calculate the elapsed tiem by subtracting timestamps.
- Handle the KeyboardInterrupt exception so ther user can press CTRL-C to quit.

### Step 1: Set Up the Program to Track Times

In [21]:
import time

# Display the program's instructions.
print('Press ENTER to begin. Afeterwards, press ENTER to "click" the stopwatch. Press Ctrl-C to quit.')
raw_input() # press Enter to begin
print('Started.')
startTime = time.time() # get the first lap's start time
lastTime = startTime
lapNum = 1

# TODO: Start tracking the lap times.

Press ENTER to begin. Afeterwards, press ENTER to "click" the stopwatch. Press Ctrl-C to quit.

Started.


### Step 2: Track and Print Lap Times

In [25]:
%%writefile stopWatch.py
import time

# Display the program's instructions.
print('Press ENTER to begin. Afeterwards, press ENTER to "click" the stopwatch.')
print('Press Ctrl-C to quit.')
raw_input() # press Enter to begin
print('Started.')
startTime = time.time() # get the first lap's start time
lastTime = startTime
lapNum = 1

# TODO: Start tracking the lap times.
try:
    while True:
        raw_input()
        lapTime = round(time.time() - lastTime, 2)
        totalTime = round(time.time() - startTime, 2)
        print('Lap #{}: {} ({})'.format(lapNum, totalTime, lapTime))
        lapNum += 1
        lastTime = time.time() # reset the last lap time
except KeyboardInterrupt:
    # Handle the Ctrl-C exception to keep its error message from displaying.
    print('\nDone.')

Overwriting stopWatch.py


In [24]:
%run stopWatch.py

Press ENTER to begin. Afeterwards, press ENTER to "click" the stopwatch. Press Ctrl-C to quit.

Started.

Lap #1: 1.49 (1.49)

Lap #2: 2.76 (1.27)

Lap #3: 3.74 (0.98)

Lap #4: 4.84 (1.1)

Lap #5: 8.38 (3.54)
quit
Lap #6: 9.31 (0.93)

Lap #7: 10.32 (1.01)

Lap #8: 11.24 (0.92)

Lap #9: 11.95 (0.71)

Lap #10: 14.12 (2.17)

Lap #11: 22.1 (7.98)

Done.


### Ideas for Similar Programs

- Create a simple timesheet app that records when you type a person's name and uses the current time to clock them in or out.
- 간단한 timesheet app을 만든다. 그것은 네가 사람의 이름을 입력할 때 기록한다. 그것은 현재 시간을 사용해서 기록한다.
- Add a feature to your program to display the elapsed time since a process started, such as a download that uses the requests module. (See Chapter 11)
- 너의 프로그램에 지나간 시간을 기록하는 기능을 추가한다. 프로그램이 실행된 후로. requests module을 사용해서 다운로드하는 프로그램 같은거
- Intermittently check how long a program has been running and offer the user a chance to cancel tasks that are taking too long.
- 간헐적으로 확인한다. 프로그램이 얼마나 실행됐고 많은 임무가 실행됐다면 임무를 취소할 수 있게 제공한다.

## The datetime Module

In [26]:
import datetime

In [27]:
datetime.datetime.now()

datetime.datetime(2015, 6, 7, 16, 16, 59, 992141)

In [36]:
dt = datetime.datetime(2015, 10, 21, 16, 29, 0)
dt

datetime.datetime(2015, 10, 21, 16, 29)

In [29]:
dt.year, dt.month, dt.day

(2015, 10, 21)

In [30]:
dt.hour, dt.minute, dt.second

(16, 29, 0)

In [32]:
datetime.datetime.fromtimestamp(1000000)

datetime.datetime(1970, 1, 12, 22, 46, 40)

In [33]:
datetime.datetime.fromtimestamp(time.time())

datetime.datetime(2015, 6, 7, 16, 19, 26, 638347)

In [34]:
dt = datetime.datetime(2015, 10, 21, 16, 29, 0, 638347)

In [35]:
dt

datetime.datetime(2015, 10, 21, 16, 29, 0, 638347)

In [38]:
# microsecond
help(datetime.datetime)

Help on class datetime in module datetime:

class datetime(date)
 |  datetime(year, month, day[, hour[, minute[, second[, microsecond[,tzinfo]]]]])
 |  
 |  The year, month and day arguments are required. tzinfo may be None, or an
 |  instance of a tzinfo subclass. The remaining arguments may be ints or longs.
 |  
 |  Method resolution order:
 |      datetime
 |      date
 |      __builtin__.object
 |  
 |  Methods defined here:
 |  
 |  __add__(...)
 |      x.__add__(y) <==> x+y
 |  
 |  __eq__(...)
 |      x.__eq__(y) <==> x==y
 |  
 |  __ge__(...)
 |      x.__ge__(y) <==> x>=y
 |  
 |  __getattribute__(...)
 |      x.__getattribute__('name') <==> x.name
 |  
 |  __gt__(...)
 |      x.__gt__(y) <==> x>y
 |  
 |  __hash__(...)
 |      x.__hash__() <==> hash(x)
 |  
 |  __le__(...)
 |      x.__le__(y) <==> x<=y
 |  
 |  __lt__(...)
 |      x.__lt__(y) <==> x<y
 |  
 |  __ne__(...)
 |      x.__ne__(y) <==> x!=y
 |  
 |  __radd__(...)
 |      x.__radd__(y) <==> y+x
 |  
 |  __reduce__(.

In [39]:
halloween2015 = datetime.datetime(2015, 10, 31, 0, 0, 0)

In [40]:
newyears2016 = datetime.datetime(2016, 1, 1, 0, 0, 0)

In [41]:
oct_31_2015 = datetime.datetime(2015, 10, 31, 0, 0, 0)

In [42]:
halloween2015 == oct_31_2015

True

In [43]:
newyears2016 > halloween2015

True

In [44]:
newyears2016 != oct_31_2015

True

### The timedelta Data Type

- duration

In [50]:
help(datetime.timedelta)

Help on class timedelta in module datetime:

class timedelta(__builtin__.object)
 |  Difference between two datetime values.
 |  
 |  Methods defined here:
 |  
 |  __abs__(...)
 |      x.__abs__() <==> abs(x)
 |  
 |  __add__(...)
 |      x.__add__(y) <==> x+y
 |  
 |  __div__(...)
 |      x.__div__(y) <==> x/y
 |  
 |  __eq__(...)
 |      x.__eq__(y) <==> x==y
 |  
 |  __floordiv__(...)
 |      x.__floordiv__(y) <==> x//y
 |  
 |  __ge__(...)
 |      x.__ge__(y) <==> x>=y
 |  
 |  __getattribute__(...)
 |      x.__getattribute__('name') <==> x.name
 |  
 |  __gt__(...)
 |      x.__gt__(y) <==> x>y
 |  
 |  __hash__(...)
 |      x.__hash__() <==> hash(x)
 |  
 |  __le__(...)
 |      x.__le__(y) <==> x<=y
 |  
 |  __lt__(...)
 |      x.__lt__(y) <==> x<y
 |  
 |  __mul__(...)
 |      x.__mul__(y) <==> x*y
 |  
 |  __ne__(...)
 |      x.__ne__(y) <==> x!=y
 |  
 |  __neg__(...)
 |      x.__neg__() <==> -x
 |  
 |  __nonzero__(...)
 |      x.__nonzero__() <==> x != 0
 |  
 |  __pos__(...

In [45]:
delta = datetime.timedelta(days=11, hours=10, minutes=9, seconds=8)

In [46]:
delta.days, delta.seconds, delta.microseconds

(11, 36548, 0)

In [47]:
delta

datetime.timedelta(11, 36548)

In [48]:
delta.total_seconds()

986948.0

In [49]:
str(delta)

'11 days, 10:09:08'

In [51]:
# weeks도 있네
delta = datetime.timedelta(weeks=1, days=11, hours=10, minutes=9, seconds=8)
delta

datetime.timedelta(18, 36548)

#### date arithmetic on datetime values

In [53]:
dt = datetime.datetime.now()
dt

datetime.datetime(2015, 6, 7, 16, 50, 36, 673428)

In [54]:
thousandDays = datetime.timedelta(days=1000)

In [55]:
dt + thousandDays

datetime.datetime(2018, 3, 3, 16, 50, 36, 673428)

In [56]:
oct21st = datetime.datetime(2015, 10, 21, 16, 29, 0)

In [57]:
aboutThirtyYears = datetime.timedelta(days=365 * 30)

In [58]:
oct21st

datetime.datetime(2015, 10, 21, 16, 29)

In [59]:
oct21st - aboutThirtyYears

datetime.datetime(1985, 10, 28, 16, 29)

In [62]:
oct21st - (2*aboutThirtyYears)

datetime.datetime(1955, 11, 5, 16, 29)

In [60]:
oct21st + aboutThirtyYears

datetime.datetime(2045, 10, 13, 16, 29)

### Pausing Until a Specific Date

In [63]:
import datetime

In [64]:
import time

In [65]:
halloween2016 = datetime.datetime(2016, 10, 31, 0, 0, 0)

In [68]:
while datetime.datetime.now() < halloween2016:
    time.sleep(1)

KeyboardInterrupt: 

### Converting datetime Objects into Strings

- strftime(): datetime object를 string으로 보여주기 위한 메소드. 
- string format time

#### Table 15-1. strftime() Directives

strftime directive | Meaning
--- | ---
%Y | Year with century, as in '2014'
%y | Year without century, '00' to '99' (1970 to 2069)
%m | Month as a decimal number, '01' to '12'
%B | Full month name, as in 'November'
%b | Abbreviated month name, as in "Nov'
%d | Day of the month, '01' to '31'
%j | Day of the year, '001' to '366'
%w | Day of the week, '0' (Sunday) to '6' (Saturday)
%A | Full weekday name, as in 'Monday'
%a | Abbreviated weekday name, as in 'Mon'
%H | Hour(24-hour clock), '00' to '23'
%I | Hour(12-hour clock), '01' to '12'
%M | Minute, '00' to '59'
%S | Second, '00' to '59'
%p | 'AM' or 'PM'
%% | Literal '%' character

In [69]:
oct21st = datetime.datetime(2015, 10, 21, 16, 29, 0)

In [70]:
oct21st.strftime('%Y/%m/%d %H:%M:%S')

'2015/10/21 16:29:00'

In [71]:
oct21st.strftime('%I:%M %p')

'04:29 PM'

In [72]:
oct21st.strftime("%B of '%y")

"October of '15"

### Converting Strings into datetime Objects

- strptime: str parse time

In [74]:
datetime.datetime.strptime('October 21, 2015', '%B %d, %Y')

datetime.datetime(2015, 10, 21, 0, 0)

In [76]:
datetime.datetime.strptime('2015/10/21 16:29:00', '%Y/%m/%d %H:%M:%S')

datetime.datetime(2015, 10, 21, 16, 29)

In [77]:
datetime.datetime.strptime("October of '15", "%B of '%y")

datetime.datetime(2015, 10, 1, 0, 0)

In [78]:
datetime.datetime.strptime("November of '63", "%B of '%y")

datetime.datetime(2063, 11, 1, 0, 0)

In [98]:
# 1970 ~ 2068
datetime.datetime.strptime("November of '68", "%B of '%y")

datetime.datetime(2068, 11, 1, 0, 0)

In [97]:
datetime.datetime.strptime("November of '69", "%B of '%y")

datetime.datetime(1969, 11, 1, 0, 0)

In [99]:
# ValueError
datetime.datetime.strptime("November of '6", "%B of '%y")

ValueError: time data "November of '6" does not match format "%B of '%y"

## Review of Python's Time Functions

- UNIX epoch timestamp: 12 AM on January 1, 1970, UTC
- A datetime object has integers
- A timedelta object time duration


In [100]:
time.time()

1433666004.44805

In [101]:
datetime.datetime.now()

datetime.datetime(2015, 6, 7, 17, 33, 30, 24238)

In [103]:
datetime.datetime.fromtimestamp(time.time())

datetime.datetime(2015, 6, 7, 17, 33, 56, 471481)

In [105]:
delta = datetime.timedelta(weeks=1, days=1, hours=10)
delta

datetime.timedelta(8, 36000)

In [106]:
delta.total_seconds()

727200.0

#### review of time functions and their parameters and return values

- The time.time() function returns an epoch timestamp float value of the current moment.
- The time.sleep(seconds) function stops the program for the amount of seconds specified by the seconds argument.
- The datetime.datetime(year, month, day, hour, minute, second) function returns a datetime object of the moment specified by the arguments. If hour, minute, or second arguments are not provided, they default to 0.
- The datetime.datetime.now() function returns a datetime object of the current moment.
- The datetime.datetime.fromtimestamp(epoch) function returns a datetime object of the moment represented by the epoch timestamp argument.
- The datetime.timedelta(weeks, days, hours, minutes, seconds, milliseconds, microseconds) function returns a timedelta object representing a duration of time. The function’s keyword arguments are all optional and do not include month or year.
- The total_seconds() method for timedelta objects returns the number of seconds the timedelta object represents.
- The strftime(format) method returns a string of the time represented by the datetime object in a custom format that’s based on the format string. See Table 15-1 for the format details.
- The datetime.datetime.strptime(time_string, format) function returns a datetime object of the moment specified by time_string, parsed using the format string argument. See Table 15-1 for the format details.

## Multithreading


In [107]:
import time, datetime

startTime = datetime.datetime(2029, 10, 31, 0, 0, 0)
while datetime.datetime.now() < startTime:
    time.sleep(1)

print('Program now starting on Halloween 2029')

KeyboardInterrupt: 

In [112]:
%%writefile test_thread.py
import threading
import time

print('Start of program.')

def takeANap():
    time.sleep(5)
    print('Wake up!')
    
threadObj = threading.Thread(target=takeANap)
threadObj.start()

print('End of program.')

Writing test_thread.py


In [113]:
!python test_thread.py

Start of program.
End of program.
Wake up!


In [114]:
!python test_thread.py

Start of program.
End of program.
Wake up!


### Passing Arguments to the Thread's Target Function

In [115]:
from __future__ import print_function

In [116]:
print('Cats', 'Dogs', 'Frogs', sep=' & ')

Cats & Dogs & Frogs


In [117]:
import threading

In [120]:
threadObj = threading.Thread(target=print, args=['Cats', 'Dogs', 'Frogs'],
                             kwargs={'sep': ' & '})
threadObj.start()

Cats & Dogs & Frogs


In [121]:
threadObj = threading.Thread(target=print('Cats', 'Dogs', 'Frogs', sep=' & '))
threadObj.start()

Cats & Dogs & Frogs


### Concurrency Issues

- 동일한 시간에 변수에 읽거나 쓰게 되면. 디버깅이 어렵게 된다.
- multiple threads로 읽거나 쓸때는 동일한 변수에 쓰지 않는다. 
- 새로운 쓰레드를 만들면 오직 그 함수의 로컬 변수만 사용한다.

## Project: Multithreaded XKCD Downloader

### Step 1: Modify the Program to Use a Function

In [123]:
%%writefile multidownloadXkcd.py
import requests
import os
import bs4
import threading
import re
import urlparse

if not os.path.exists('xkcd'):
    os.makedirs('xkcd')
    
def downloadXkcd(startComic, endComic):
    for urlNumber in range(startComic, endComic):
        # Download the page.
        print('Downloading page http://xkcd.com/{}...'.format(urlNumber))
        res = requests.get('http://xkcd.com/{}'.format(urlNumber))
        res.raise_for_status()
        
        soup = bs4.BeautifulSoup(res.text)
        
        # Find the URL of the comic image.
        comicElem = soup.select('#comic img')
        if comicElem == []:
            print('Could not find comic image.')
        else:
            comicUrl = comicElem[0].get('src')
        if comicUrl.startswith('/') or not 'http' in comicUrl:
            comicUrl = urlparse.urljoin('http://xkcd.com', comicUrl)
            
        # TODO: Download the image.
        print('Downloading image %s...' % (comicUrl))
        res = requests.get(comicUrl)
        res.raise_for_status()
    
        # TODO: Save the image to ./xkcd.
        with open(os.path.join('xkcd', os.path.basename(comicUrl)), 'wb') as imageFile:
            for chunk in res.iter_content(100000):
                imageFile.write(chunk)
        
    # TODO: Create and start the Thread objects.
    downloadThreads = [] # a list of all the Thread objects
    for i in range(0, 1600, 100): # loops 14 times, creates 14 threads
        downloadThread = threading.Thread(target=downloadXkcd, args=(i, i+99))
        downloadThreads.append(downloadThread)
        downloadThread.start()
    # TODO: Wait for all threads to end.

Writing multidownloadXkcd.py


### Step 3: Wait for All Threads to End

In [143]:
!rm -rf xkcd

#### requests.get 오류

- 프로그램이 실행하다가 제대로 파일을 받아오지 못해서 오류가 난다.
- 오류가 난 것을 저장해 놓고 모두 끝난 다음에 다시 그것들만 받아오기? 그래도 또 오류가 발생하면?
- 그렇다면 받아온 것을 저장해 놓는건 어때? 받아놓은 것을 모두 적어놓고 검사하면 되겠네

In [149]:
%%writefile multidownloadXkcd.py
import requests
import os
import bs4
import threading
import re
import urlparse

if not os.path.exists('xkcd'):
    os.makedirs('xkcd')
    
# def write_wrong():
    

def downloadXkcd(startComic, endComic):
    for urlNumber in range(startComic, endComic):
        # Download the page.
        print('Downloading page http://xkcd.com/{}...'.format(urlNumber))
        res = requests.get('http://xkcd.com/{}'.format(urlNumber))
        res.raise_for_status()
        
        soup = bs4.BeautifulSoup(res.text)
        
        # Find the URL of the comic image.
        comicElem = soup.select('#comic img')
        if comicElem == []:
            print('Could not find comic image.')
        else:
            comicUrl = comicElem[0].get('src')
        if comicUrl.startswith('/') or not 'http' in comicUrl:
            comicUrl = urlparse.urljoin('http://xkcd.com', comicUrl)
            
        # TODO: Download the image.
        print('Downloading image %s...' % (comicUrl))
        try:
            res = requests.get(comicUrl)
            res.raise_for_status()
        except Exception as e:
            print(e)
    
        # TODO: Save the image to ./xkcd.
        with open(os.path.join('xkcd', os.path.basename(comicUrl)), 'wb') as imageFile:
            for chunk in res.iter_content(100000):
                imageFile.write(chunk)
                
def main():
    # TODO: Create and start the Thread objects.
    downloadThreads = [] # a list of all the Thread objects
    for i in range(0, 1600, 100): # loops 14 times, creates 14 threads
        # i+100 is correct. not i+99
        downloadThread = threading.Thread(target=downloadXkcd, args=(i, i+100))
        downloadThreads.append(downloadThread)
        downloadThread.start()
    # TODO: Wait for all threads to end.
    for downloadThread in downloadThreads:
        downloadThread.join()
    print('Done.')
    
if __name__ == '__main__':
    main()

Overwriting multidownloadXkcd.py


In [142]:
# for i in range(0, 99):
#     print(i)

In [150]:
!python multidownloadXkcd.py


Exception in thread Thread-1:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "multidownloadXkcd.py", line 16, in downloadXkcd
    res.raise_for_status()
  File "/Users/re4lfl0w/.virtualenvs/general/lib/python2.7/site-packages/requests/models.py", line 831, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
HTTPError: 404 Client Error: Not Found

Exception in thread Thread-5:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*s

In [145]:
!ls -l xkcd | wc -l

    1165


In [151]:
!ls -l xkcd | wc -l

    1337


## Launching Other Programs from Python

- Python으로 다른 프로그램을 실행할 수 있다. Popen(Process open)을 사용하면 된다.

In [152]:
!open /Applications/Calculator.app/Contents/MacOS/Calculator

In [153]:
import subprocess

In [154]:
subprocess.Popen('/Applications/Calculator.app/Contents/MacOS/Calculator')

In [155]:
subprocess.Popen('/Applications/Calculator.app/Contents/MacOS/Calculator')

In [156]:
subprocess.Popen(['open', '/Applications/Calculator.app/'])

In [162]:
subprocess.Popen(['open', './census2010.py'])

#### 유용한 poll() 과 wait()

- poll(): 실행하는 코드가 끝났는지 친구에게 물어본다. 아직 프로그램이 실행하고 있을 때 poll()을 호출하게 되면 None을 돌려준다.
  - 프로그램을 종료하게 되면 integer exit code를 돌려준다.
  - exit code of 0: 에러 x. 에러 여부를 알려준다.
  - 보통 1이면 에러 발생. 그러나 프로그램들에 따라서 바뀐다.
- wait(): 프로그램이 종료될 때까지 기다린다. wait()의 리턴값은 프로세스의 integer exit code다.
- [python \- Difference between Popen.poll() and Popen.wait() \- Stack Overflow](http://stackoverflow.com/questions/13832734/difference-between-popen-poll-and-popen-wait)
- [python \- Using subprocess wait() and poll() \- Stack Overflow](http://stackoverflow.com/questions/2995983/using-subprocess-wait-and-poll)


In [20]:
calcProc = subprocess.Popen(['open', '/Applications/Calculator.app/'])

In [21]:
calcProc.poll()

0

In [22]:
calcProc.poll() == None

False

In [23]:
calcProc.wait()

0

In [24]:
calcProc.poll()

0

In [1]:
import time
import subprocess

p = subprocess.Popen("sleep 30", shell=True)
# Better: p = subprocess.Popen(["sleep", "30"])

# Wait until process terminates
while p.poll() is None:
    time.sleep(0.5)

# It's done
print("Process ended, ret code:", p.returncode)

('Process ended, ret code:', 0)


In [4]:
p = subprocess.Popen("sleep 10", shell=True)
while p.poll() is None:
    print("Still working...{}".format(p.poll()))
    time.sleep(0.5)
print('Done.')
print(p.poll())

Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Still working...None
Done.
0


In [7]:
p = subprocess.Popen("sleep 10", shell=True)
while p.wait() is None:
    print("Still working...")
    time.sleep(0.5)

In [10]:
# wait는 sleep 10이 끝날때까지 기다린 후에 결과값으로 0을 돌려주네.
# 완전 block 이잖아..?
p = subprocess.Popen("sleep 10", shell=True)
p.wait()

0

### Passing Command Line Arguments to Popen()

### Mac os X Task Scheduler

- [10 Apps \- iPhone 실전 프로젝트 따라하기 :: launchd를 사용해서 원하는 서비스 실행시키기](http://10apps.tistory.com/116)
- [Tech Blog :: Mac에서 정해진 시간에 자동으로 프로그램 실행하는 법](http://techblog.tistory.com/entry/launchd)
- [Lingon \- Peter Borg Apps](https://www.peterborgapps.com/lingon/): 유료 버전으로 구매를 해야되서 보류
- [soma\-zone: LaunchControl](http://www.soma-zone.com/LaunchControl/): 이게 무료로 쓸 수 있는 것인듯. GUI도 지원

In [14]:
import sys
import plistlib

root = {}
root['Label'] = 'com.jinni.ipfwd.ftp'
root['ProgramArguments'] = '/sbin/ipfw add 100 forward 127.0.0.1,2121 ip from any to any 21 in'.split()
root['KeepAlive'] = False
root['RunAtLoad'] = True

print plistlib.writePlist(root, sys.stdout)

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>KeepAlive</key>
	<false/>
	<key>Label</key>
	<string>com.jinni.ipfwd.ftp</string>
	<key>ProgramArguments</key>
	<array>
		<string>/sbin/ipfw</string>
		<string>add</string>
		<string>100</string>
		<string>forward</string>
		<string>127.0.0.1,2121</string>
		<string>ip</string>
		<string>from</string>
		<string>any</string>
		<string>to</string>
		<string>any</string>
		<string>21</string>
		<string>in</string>
	</array>
	<key>RunAtLoad</key>
	<true/>
</dict>
</plist>
None


In [13]:
d = subprocess.Popen(['python', 'census2010.py'])
d

### Opening Files with Default Applications

- depending on your system

In [38]:
import platform

In [41]:
platform.system()

'Darwin'

In [53]:
platform.architecture()

('64bit', '')

In [51]:
int(platform.python_version_tuple()[0])

2

In [46]:
platform.machine()

'x86_64'

In [68]:
%%writefile tmp/test_hello.py
fileObj = open('hello.txt', 'w')
fileObj.write('Hello world!')
fileObj.close()

import subprocess
import platform

plat = platform.system()

if plat == 'Darwin':
    subprocess.Popen(['open', 'hello.txt'])
elif plat == 'Linux':
    subprocess.Popen(['open', 'hello.txt'])
elif plat == 'Windows':
    subprocess.Popen(['start', 'hello.txt'])

Overwriting tmp/test_hello.py


- [cross platform \- How can I find the current OS in Python? \- Stack Overflow](http://stackoverflow.com/questions/110362/how-can-i-find-the-current-os-in-python): platform 사용해서 구분

In [70]:
import platform
import sys

def linux_distribution():
    try:
        return platform.linux_distribution()
    except:
        return "N/A"

print("""Python version: %s
dist: %s
linux_distribution: %s
system: %s
machine: %s
platform: %s
uname: %s
version: %s
mac_ver: %s
""" % (
sys.version.split('\n'),
str(platform.dist()),
linux_distribution(),
platform.system(),
platform.machine(),
platform.platform(),
platform.uname(),
platform.version(),
platform.mac_ver(),
))

Python version: ['2.7.6 (default, Sep  9 2014, 15:04:36) ', '[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)]']
dist: ('', '', '')
linux_distribution: ('', '', '')
system: Darwin
machine: x86_64
platform: Darwin-14.3.0-x86_64-i386-64bit
uname: ('Darwin', 're4lfl0wui-MacBook-Pro.local', '14.3.0', 'Darwin Kernel Version 14.3.0: Mon Mar 23 11:59:05 PDT 2015; root:xnu-2782.20.48~5/RELEASE_X86_64', 'x86_64', 'i386')
version: Darwin Kernel Version 14.3.0: Mon Mar 23 11:59:05 PDT 2015; root:xnu-2782.20.48~5/RELEASE_X86_64
mac_ver: ('10.10.3', ('', '', ''), 'x86_64')



## Project: Simple Countdown Program

- 간단한 stopwatch application을 찾는건 어렵다.
- countdown 프로그램의 끝에 알람이 울리게 한다.

### High Level Logic

- Count down from 60
- Play a sound file(alarm.wav) when the countdown reaches zero.

### Code Level Logic

- 1개의 숫자가 줄어들 때마다 time.sleep() 함수를 사용한다.
- subprocess.Popen() 으로 sound file을 default application으로 실행한다.

### Step 1: Count Down

In [73]:
from __future__ import print_function

In [74]:
import time
import subprocess

timeLeft = 60
while timeLeft > 0:
    print(timeLeft, end='')
    time.sleep(1)
    timeLeft = timeLeft - 1
    
# TODO: At the end of the countdown, play a sound file.

6059585756555453

KeyboardInterrupt: 

### Step 2: Play the Sound File

In [77]:
import time
import subprocess

timeLeft = 10
while timeLeft > 0:
    print(timeLeft, end=' ')
    time.sleep(1)
    timeLeft = timeLeft - 1
    
subprocess.Popen(['open', 'src/alarm.wav'])

10 9 8 7 6 5 4 3 2 1 

### Ideas for Similar Programs

- time.sleep() 을 사용한다. 유저에게 이 행동을 취소할 수 있는 기회를 준다. ctrl-C를 눌러서
- KeyboardInterrupt exceptions은 try와 except 문장을 사용해서 해라
- 긴시간을 카운트다운 한다면 timedelta를 사용해서 해라. days, hours, minutes, and seconds

## Summary

- Unix epoch(January 1, 1970, at midnight, UTC)
- time.sleep()은 몇 초간 행동을 멈추게 한다.
- 그러나 프로그램을 스케쥴에 맞게끔 실행시키려면 네 운영체제에 맞는 스케줄러를 사용해라.
- threading module은 여러개의 threads를 만들 수 있다. 여러개의 파일을 동시에 다운로드 받는데 좋다. 오직 로컬 변수만 사용해라. 동시성 문제가 발생할 수 있다.
- subprocess.Popen() 을 사용하면 커맨드라인으로 몇개의 인자를 넘길 수 있다. 
- start, open 등을 사용하면 운영체제에 기본 설정되어있는 애플리케이션으로 실행한다.

## Practice Projects

### Prettfied Stopwatch

- rjust(), ljust()
- [2.13 텍스트 정렬 - Python Cookbook](http://nbviewer.ipython.org/github/re4lfl0w/ipython/blob/master/books/python_cookbook/ch02_02.ipynb)

In [268]:
import pyperclip

In [270]:
a = 10

In [271]:
pyperclip.copy(a)

In [274]:
%%writefile tmp/stopWatch_pretty.py
import time
import pyperclip

# Display the program's instructions.
print('Press ENTER to begin. Afeterwards, press ENTER to "click" the stopwatch.')
print('Press Ctrl-C to quit.')
raw_input() # press Enter to begin
print('Started.')
startTime = time.time() # get the first lap's start time
lastTime = startTime
lapNum = 1

# TODO: Start tracking the lap times.
try:
    while True:
        raw_input()
        lapTime = round(time.time() - lastTime, 2)
        totalTime = round(time.time() - startTime, 2)
        print('Lap #{:>5}: {:>7} ({:>7})'.format(lapNum, 
                                                   '{:>.2f}'.format(totalTime),
                                                   '{:>.2f}'.format(lapTime)))
        lapNum += 1
        lastTime = time.time() # reset the last lap time
        pyperclip.copy(totalTime)
except KeyboardInterrupt:
    # Handle the Ctrl-C exception to keep its error message from displaying.
    print('\nDone.')

Overwriting tmp/stopWatch_pretty.py


In [275]:
%run tmp/stopWatch_pretty.py

Press ENTER to begin. Afeterwards, press ENTER to "click" the stopwatch.
Press Ctrl-C to quit.

Started.

Lap #    1:    0.58 (   0.58)

Lap #    2:    3.37 (   2.79)

Lap #    3:   10.46 (   7.09)

Lap #    4:   19.80 (   9.34)

Done.


- 19.8

### Scheduled Web Comic Downloader

- 마지막 방문 후에 업데이트가 되었다면 여러개의 웹 코믹을 체크하고 자동으로 다운로드 받는다.
- 그렇게 되면 업데이트가 되었는지 계속 확인되는 것에서 자유로워질 수 있다.